### DISCUSSION TITLE
When free bio

### DISCUSSION THREAD

**Discussion by yuanfangcc** [openzfs/zfs#15173] [Q&A] [answered]
In this funtion,bio is not free.
`static int
vdev_disk_io_flush(struct block_device *bdev, zio_t *zio)
{
	struct request_queue *q;
	struct bio *bio;

	q = bdev_get_queue(bdev);
	if (!q)
		return (SET_ERROR(ENXIO));

	bio = vdev_bio_alloc(bdev, GFP_NOIO, 0);
	if (unlikely(bio == NULL))
		return (SET_ERROR(ENOMEM));

	bio->bi_end_io = vdev_disk_io_flush_completion;
	bio->bi_private = zio;
	bio_set_flush(bio);
	vdev_submit_bio(bio);
	invalidate_bdev(bdev); 
	
	if (bio != NULL)
		printk(KERN_INFO "=========vdev_disk_io_flush====bio is NOT free==\n");

	return (0);
}`

This is printed.I don't understand, when is bio released.

**Comment by robn** [comment, marked as answer]
All you're testing is that the pointer isn't NULL, which says nothing about what, if anything, its pointing to.

As it is, the bio is handed to the kernel (`vdev_submit_bio_impl` -> `submit_bio`), which then takes ownership of it.

When the IO operation is complete, the kernel calls `vdev_disk_io_flush_completion` with the completed bio. This usually the bio that was passed in, but not necessarily - the kernel may modify it, split it or whatever it wants. Regardless, `bio_put(bio)` returns the bio to the kernel when we're done with it, and it will free it or reuse it as it likes.

**Reply by yuanfangcc** [reply]
Got it. Thank you.
